### PR TITLE
Add unit testing framework and tests for existing code (and assorted fixes and breaking API changes)

### PR DESCRIPTION
**Pull request by Sophon96**
Using catch2 and integration with meson's testing harness (so tests can be run with `meson test`).

### CODE REVIEW

**Comment by Sophon96** [comment]
Incomplete, blocked for test requirements from @adamhutchings for the matrix and vector classes.

**Comment by adamhutchings** [comment]
Are we merging this after #25?

**Comment by Sophon96** [comment]
> Are we merging this after #25?

No, I need to write tests for the existing code as well (though this is blocked by #25 as well).

File: tests/math/matrix.cpp
Comment (adamhutchings) [inline tests/math/matrix.cpp]: Newline?

**Comment by adamhutchings** [comment]
Is development on this able to proceed or are we still waiting on something else?

**Comment by Sophon96** [comment]
> Is development on this able to proceed or are we still waiting on something else?

Waiting on nothing, just time and writing tests. I think it would be interesting if we want to do test-driven development or behavior-driven development for v1 (i.e. draft the API, write all the tests, and once all tests pass v1 is complete).

**Comment by adamhutchings** [comment]
I think that is a worthwhile discussion to have, but if we have the basis of testing set up, we can probably close this PR and add more tests as they come (or however else we agree to do it).

**Comment by Sophon96** [comment]
Blocked #38

**Comment by Sophon96** [comment]
~~@adamhutchings @JakeRoggenbuck Can any of y'all find where I messed up the code in the test? `tests/core/test_matrix.cpp`~~

There are numerous bugs I've found in the actual lib source code. I'm working on fixing them right now.

**Comment by Sophon96** [comment]
(this should be squashed as well, this git history is unimaginably bad)

**Comment by Sophon96** [comment]
TODO before this can be merged
- [x] Update docs (just README.md for now)
- [x] Tests for the rest of the code (Model, though not the entirety because of the lack of actual implementation; getters and setters)
- [x] Add more getters (Matrix needs some, Model too)
- [ ] Review refactored code and breaking API changes introduced in this PR.

**Comment by Sophon96** [comment]
C API is not tested.

**Comment by Sophon96** [comment]
oh wait i need to fix ci again don't merge

**Comment by Sophon96** [comment]
@adamhutchings @JakeRoggenbuck Ready to review

File: tests/core/math/test_activation_functions.cpp
Comment (adamhutchings) [inline tests/core/math/test_activation_functions.cpp:36]: I don't think this is UB because we can just arbitrarily choose a value for the derivative. Like you chose to do, probably better not to test this because the behavior of the derivative at the discontinuity doesn't matter.

**Review by adamhutchings — APPROVED** [review]
Much appreciated!

**Review by JakeRoggenbuck — APPROVED** [review]

**Pull request merged**
